File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/generate/ExtensionsHelper.java
package io.github.potjerodekool.openapi.common.generate;

import java.util.List;
import java.util.Map;

public final class ExtensionsHelper {

    private ExtensionsHelper() {
    }

    public static <T> T getExtension(final Map<String, Object> extensions, final String name, final Class<T> resultClass) {
        if (extensions == null || extensions.isEmpty()) {
            return getDefaultValue(resultClass);
        }

        final var extension = extensions.get(name);

        if (extension == null) {
            return getDefaultValue(resultClass);
        }

        return (T) extension;
    }

    private static <T> T getDefaultValue(final Class<T> resultClass) {
        if (resultClass == List.class) {
            return (T) List.of();
        } else {
            return null;
        }
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/generate/ExtensionsHelper.kt
package io.github.potjerodekool.openapi.common.generate

import io.github.potjerodekool.openapi.common.generate.annotation.TypeInfo
import java.util.function.Supplier

object ExtensionsHelper {

    @JvmStatic
    fun <T> getExtension(
        extensions: MutableMap<String, Any?>?,
        name: String?,
        resultClass: Class<T>
    ): T? {
        if (extensions == null || extensions.isEmpty()) {
            return getDefaultValue(resultClass)
        }

        val extension = extensions[name] ?: return getDefaultValue(resultClass)

        return extension as T
    }

    fun <T> getExtension(
        extensions: MutableMap<String, Any?>?,
        name: String?,
        returnTypeInfo: TypeInfo<T>?
    ): T? {
        return getExtension(extensions, name, returnTypeInfo, null)
    }

    fun <T> getExtension(
        extensions: MutableMap<String, Any?>?,
        name: String?,
        returnTypeInfo: TypeInfo<T>?,
        defaultValueSupplier: Supplier<T>?
    ): T? {
        if (extensions == null
            || extensions.isEmpty()
        ) {
            return defaultValueSupplier?.get()
        }

        val extension = extensions[name]
            ?: return defaultValueSupplier?.get()

        return extension as T
    }

    private fun <T> getDefaultValue(resultClass: Class<T>): T? {
        return if (resultClass == MutableList::class.java) {
            listOf<Any>() as T
        } else {
            null
        }
    }
}
